drivers/SmartThings/zigbee-thermostat/src/vimar/init.lua

```diff
@@ -0,0 +1,213 @@
+    Thermostat.attributes.ControlSequenceOfOperation,
+    Thermostat.attributes.ThermostatRunningState,
+    Thermostat.attributes.SystemMode
+  }
+
+  local vimar_thermostat_profile = device:get_field(VIMAR_CURRENT_PROFILE)
+  if vimar_thermostat_profile == ThermostatMode.thermostatMode.heat.NAME then
+    attributes[#attributes + 1] = Thermostat.attributes.OccupiedHeatingSetpoint
+  elseif vimar_thermostat_profile == ThermostatMode.thermostatMode.cool.NAME then
+    attributes[#attributes + 1] = Thermostat.attributes.OccupiedCoolingSetpoint
+  end
+
+  for _, attribute in pairs(attributes) do
+    device:send(attribute:read(device))
+  end
+end
+
+-- NOTE: Whenever the physical button for the current mode is pressed on the device, this function changes the device profile.
+--       If the Thermostat is OFF, no profile change is required.
+local vimar_thermostat_mode_handler = function(driver, device, thermostat_mode)
+  local mode = VIMAR_THERMOSTAT_MODE_MAP[thermostat_mode.value].NAME
+  -- If is a known supported mode, then apply the change
+  if VIMAR_THERMOSTAT_MODE_MAP[thermostat_mode.value] then
+    local vimar_thermostat_profile = device:get_field(VIMAR_CURRENT_PROFILE)
+    -- HEAT: if the previous mode was cool, update profile
+    if mode == ThermostatMode.thermostatMode.heat.NAME then
+      if mode ~= vimar_thermostat_profile then
+        device:try_update_metadata({ profile = VIMAR_THERMOSTAT_HEATING_PROFILE })
+        device:set_field(VIMAR_CURRENT_PROFILE, mode)
+        device.thread:call_with_delay(2, function(d)
+          vimar_thermostat_do_refresh(driver, device)
+        end)
+      end
+      -- COOL: if the previous mode was heat, update profile
+    elseif mode == ThermostatMode.thermostatMode.cool.NAME then
+      if mode ~= vimar_thermostat_profile then
+        device:try_update_metadata({ profile = VIMAR_THERMOSTAT_COOLING_PROFILE })
+        device:set_field(VIMAR_CURRENT_PROFILE, mode)
+        device.thread:call_with_delay(2, function(d)
+          vimar_thermostat_do_refresh(driver, device)
+        end)
+      end
+    end
+    device:emit_event(VIMAR_THERMOSTAT_MODE_MAP[thermostat_mode.value]())
+  end
+end
+
+-- NOTE: override default thermostat mode map; logic is the same from the original driver
+local vimar_set_thermostat_mode = function(driver, device, command)
+  for zigbee_attr_val, st_cap_val in pairs(VIMAR_THERMOSTAT_MODE_MAP) do
+    if command.args.mode == st_cap_val.NAME then
+      device:send_to_component(command.component, Thermostat.attributes.SystemMode:write(device, zigbee_attr_val))
+      device.thread:call_with_delay(1, function(d)
+        device:send_to_component(command.component, Thermostat.attributes.SystemMode:read(device))
+      end)
+      break
+    end
+  end
+end
+
+
+local function vimar_thermostat_heating_setpoint_handler(driver, device, value, zb_rx)
+  local setpoint = value.value / 100.0
+  local unit_type = "C"
+  if (setpoint >= 41.0) then
```

Comment:
as far as I know, the zigbee spec requires these reports to be in Celsius

you can just leave these handlers as the defaults

Reply:
here's the source from the default implementation:

```
function ths_defaults.thermostat_heating_setpoint_handler(driver, device, value, zb_rx)
  local raw_temp = value.value
  local celc_temp = raw_temp / 100.0
  local temp_scale = "C"
  device:emit_event_for_endpoint(zb_rx.address_header.src_endpoint.value, capabilities.thermostatHeatingSetpoint.heatingSetpoint({value = celc_temp, unit = temp_scale }))
end
```

I believe there's some code in the hub that rounds to tenths of a degree, but that would be hit regardless of whether you're using the default or not.